Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp

```diff
@@ -1343,28 +1429,95 @@ ExceptionOr<GStreamerMediaEndpoint::Backends> GStreamerMediaEndpoint::createTran
-            g_value_unset(&value);
+            gst_value_list_append_and_take_value(&encodingsValue, &value);
         }
-    } else {
-        for (auto& encoding : init.sendEncodings) {
-            auto encodingData = fromRTCEncodingParameters(encoding);
+    } else if (!init.sendEncodings.isEmpty()) {
+        auto sendEncodings = init.sendEncodings;
+        if (init.sendEncodings.size() > 10) {
+            GST_WARNING_OBJECT(m_pipeline.get(), "Too many (%zu) sendEncodings requested for video transceiver. Limiting to 10.", init.sendEncodings.size());
+            sendEncodings = sendEncodings.subvector(0, 10);
+        }
+        Vector<String> allRids;
+        Vector<double> scaleValues;
+        scaleValues.reserveInitialCapacity(sendEncodings.size());
+        if (sendEncodings.size() == 1 && sendEncodings[0].scaleResolutionDownBy)
+            scaleValues.append(sendEncodings[0].scaleResolutionDownBy.value());
+        else if (allOf(sendEncodings, [](auto& encoding) { return encoding.scaleResolutionDownBy.value_or(1) == 1; })) {
+            for (unsigned i = sendEncodings.size() - 1; i >= 1; i--)
+                scaleValues.append(i * 2);
+            scaleValues.append(1);
+        }
+        for (unsigned i = 0; i < sendEncodings.size(); i++) {
+            auto& encoding = sendEncodings[i];
+            if (allRids.contains(encoding.rid))
+                return Exception { ExceptionCode::TypeError, makeString("Duplicate rid:"_s, encoding.rid) };
+            allRids.append(encoding.rid);
+
+            auto encodingCopy = encoding;
+            if (i < scaleValues.size())
+                encodingCopy.scaleResolutionDownBy = scaleValues[i];
+            auto encodingData = fromRTCEncodingParameters(encodingCopy, kind);
+            if (encodingData.hasException())
+                return encodingData.releaseException();
+
             GValue value = G_VALUE_INIT;
             g_value_init(&value, GST_TYPE_STRUCTURE);
-            gst_value_set_structure(&value, encodingData.get());
-            gst_value_list_append_value(&encodingsValue, &value);
-            g_value_unset(&value);
+            gst_value_set_structure(&value, encodingData.returnValue().get());
+            gst_value_list_append_and_take_value(&encodingsValue, &value);
         }
+        if (allRids.isEmpty() && sendEncodings.size() > 1)
+            return Exception { ExceptionCode::TypeError, "Missing rid"_s };
+        if (allRids.size() > 1 && anyOf(allRids, [](auto& rid) { return rid.isNull() || rid.isEmpty(); }))
+            return Exception { ExceptionCode::TypeError, "Empty rid"_s };
+        if (allRids.size() == 1 && allRids[0] == emptyString())
+            return Exception { ExceptionCode::TypeError, "Empty rid"_s };
+    } else {
+        GUniquePtr<GstStructure> encodingData(gst_structure_new("encoding-parameters", "encoding-name", G_TYPE_STRING, "H264", "payload", G_TYPE_INT, 97, "active", G_TYPE_BOOLEAN, TRUE, nullptr));
```

Comment:
Why H264 and payload 97 hardcoded here?

Reply:
https://datatracker.ietf.org/doc/html/rfc7742#section-6

We could use VP8 too, not sure that's better.

Reply:
Hmm iiuc, perhaps VP8 would be better when it gets shipped in distros without proprietary codecs by default.

Reply:
Well now with openh264 distros are supposed to have some (but bad) h.264 support, or is that a Fedora-only feature?

Anyway, I guess we can check here if h264 is supported and fallback to vp8.

Reply:
Btw I don't think this is a blocker for this PR. It can also be done in a follow-up if you prefer.

Reply:
```diff
diff --git a/Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp b/Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp
index 2ec98f8fd98c..6f9c60f7de5c 100644
--- a/Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp
+++ b/Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp
@@ -1510,11 +1510,23 @@ ExceptionOr<GStreamerMediaEndpoint::Backends> GStreamerMediaEndpoint::createTran
         if (allRids.size() == 1 && allRids[0] == emptyString())
             return Exception { ExceptionCode::TypeError, "Empty rid"_s };
     } else {
-        GUniquePtr<GstStructure> encodingData(gst_structure_new("encoding-parameters", "encoding-name", G_TYPE_STRING, "H264", "payload", G_TYPE_INT, 97, "active", G_TYPE_BOOLEAN, TRUE, nullptr));
-        GValue value = G_VALUE_INIT;
-        g_value_init(&value, GST_TYPE_STRUCTURE);
-        gst_value_set_structure(&value, encodingData.get());
-        gst_value_list_append_and_take_value(&encodingsValue, &value);
+        String fallbackCodec = emptyString();
+        for (auto& codec : codecs) {
+            if (codec.mimeType == "video/H264"_s || codec.mimeType == "video/VP8"_s) {
+                fallbackCodec = codec.mimeType.substring(6);
+                break;
+            }
+        }
+
+        GST_DEBUG_OBJECT(m_pipeline.get(), "Fallback codec: %s", fallbackCodec.ascii().data());
+        if (!fallbackCodec.isEmpty()) {
+            GUniquePtr<GstStructure> encodingData(gst_structure_new("encoding-parameters", "encoding-name", G_TYPE_STRING, fallbackCodec.ascii().data(), "payload", G_TYPE_INT, 97, "active", G_TYPE_BOOLEAN, TRUE, nullptr));
+            GValue value = G_VALUE_INIT;
+            g_value_init(&value, GST_TYPE_STRUCTURE);
+            gst_value_set_structure(&value, encodingData.get());
+            gst_value_list_append_and_take_value(&encodingsValue, &value);
+        } else
+            WTFLogAlways("Missing video encoder / RTP payloader. Please install an H.264 encoder and/or a VP8 encoder");
     }
 
     gst_structure_set_value(initData.get(), "encodings", &encodingsValue);
```